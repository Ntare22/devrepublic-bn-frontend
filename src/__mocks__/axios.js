export default {
  post: jest.fn(() => Promise.resolve({ data: { } })),
  get: jest.fn(() => Promise.resolve({ data: { data: { } } })),
  patch: jest.fn(() => Promise.resolve({ data: { data: { } } })),
};
