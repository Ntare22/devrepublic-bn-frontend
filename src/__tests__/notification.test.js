import React from 'react';
import axios from 'axios';
import dotenv from 'dotenv';
import { mount } from 'enzyme';
import io from 'socket.io-client';
import renderer from 'react-test-renderer';
import { IntlProvider } from 'react-intl';
import Notification from '../components/notification/Notifcation';
import translation from '../components/languages/en.json';

dotenv.config();
jest.mock('socket.io-client', () => {
  const emit = jest.fn();
  const on = jest.fn(() => [{
    id: 'fafwefwefwefw', content: 'this is a comment', createdAt: '2019-02-02', status: 'read',
  }]);
  const socket = { emit, on, connect: jest.fn() };
  return jest.fn(() => socket);
});

describe('<Notification>', () => {
  beforeEach(() => {
    io.mockClear();
    io().on.mockClear();
    io().emit.mockClear();
  });
  it('render correctly date component', () => {
    const container = (
      <IntlProvider defaultLocale="en" locale="en" messages={translation}>
        <Notification token="thisisatoken" />
      </IntlProvider>
    );
    const DateInputComponent = renderer.create(container).toJSON();
    expect(DateInputComponent).toMatchSnapshot();
  });
  it('sends data on login and subscribes for response', () => {
    const container = (
      <IntlProvider defaultLocale="en" locale="en" messages={translation}>
        <Notification token="thisisatoken" />
      </IntlProvider>
    );
    mount(container);
    expect(io).toHaveBeenCalledWith(process.env.BACKEND_LINK, { query: { token: 'thisisatoken' } });
  });
  it('should have default state', () => {
    const container = (
      <IntlProvider defaultLocale="en" locale="en" messages={translation}>
        <Notification token="thisisatoken" />
      </IntlProvider>
    );
    const wrapper = mount(container);
    const defaultState = wrapper.childAt(0).instance().state;
    expect(defaultState).toEqual({
      show: false, showMore: false, notifications: [], error: false,
    });
  });
  it('should have default state', () => {
    const container = (
      <IntlProvider defaultLocale="en" locale="en" messages={translation}>
        <Notification token="thisisatoken" />
      </IntlProvider>
    );
    const wrapper = mount(container).childAt(0);
    wrapper.setState({
      notifications: [{
        id: 'fafwefwefwefw', content: 'this is a comment', createdAt: '2019-02-02', status: 'unread',
      }],
    });
    wrapper.find('.notification-item');
    const { state } = wrapper.instance();
    expect(state.notifications).toEqual([{
      id: 'fafwefwefwefw', content: 'this is a comment', createdAt: '2019-02-02', status: 'unread',
    }]);
  });
  it('should show the notification panel when the icon is click', () => {
    const container = (
      <IntlProvider defaultLocale="en" locale="en" messages={translation}>
        <Notification token="thisisatoken" />
      </IntlProvider>
    );
    const wrapper = mount(container).childAt(0);
    const bell = wrapper.find('.notification-bell');
    const notificationPanel = wrapper.find('.notif-panel');
    bell.simulate('click');
    const { state } = wrapper.instance();
    expect(state.show).toBe(true);
    expect(notificationPanel.instance().style.height).toBe('280px');
  });
  it('should show more notification when view more button is clicked', () => {
    const container = (
      <IntlProvider defaultLocale="en" locale="en" messages={translation}>
        <Notification token="thisisatoken" />
      </IntlProvider>
    );
    const wrapper = mount(container).childAt(0);
    const viewMore = wrapper.find('#view-more');
    const notificationPanel = wrapper.find('.notif-panel');
    viewMore.simulate('click');
    const { state } = wrapper.instance();
    expect(state.showMore).toBe(true);
    expect(notificationPanel.instance().style.height).toBe('440px');
  });
  it('should show an error message if all notifications are already seen', () => {
    const container = (
      <IntlProvider defaultLocale="en" locale="en" messages={translation}>
        <Notification token="thisisatoken" />
      </IntlProvider>
    );
    const wrapper = mount(container).childAt(0);
    const errorMessage = wrapper.find('p span');
    wrapper.setState({ error: true });
    expect(errorMessage.instance().style.display).toBe('block');
  });
  it('should sent all notifications as read', () => {
    axios
      .patch.mockImplementationOnce(() => Promise.resolve({ response: { data: { status: 200 } } }));
    const container = (
      <IntlProvider defaultLocale="en" locale="en" messages={translation}>
        <Notification token="thisisatoken" />
      </IntlProvider>
    );
    const wrapper = mount(container).childAt(0);
    const readAll = wrapper.find('#mark-all');
    readAll.simulate('click');
  });
  it('should show error message when user has read all notif but click on the read all button', () => {
    axios
      .patch
      .mockImplementationOnce(
        () => Promise.reject(new Error({ response: { data: { status: 200 } } })),
      );
    const container = (
      <IntlProvider defaultLocale="en" locale="en" messages={translation}>
        <Notification token="thisisatoken" />
      </IntlProvider>
    );
    const wrapper = mount(container).childAt(0);
    const readAll = wrapper.find('#mark-all');
    readAll.simulate('click');
  });
  it('should show an error message if all notifications are already seen', () => {
    const container = (
      <IntlProvider defaultLocale="en" locale="en" messages={translation}>
        <Notification token="thisisatoken" />
      </IntlProvider>
    );
    const wrapper = mount(container).childAt(0);
    wrapper.instance().setNotification(JSON.stringify({
      notif: [{
        id: 'wewdwd', content: 'this is another comment', status: 'read', createdAt: '2019-04-04',
      }],
    }));
    const { state } = wrapper.instance();
    expect(state.notifications).toEqual([{
      id: 'wewdwd', content: 'this is another comment', status: 'read', createdAt: '2019-04-04',
    }]);
  });
  it('test if new notifications are added to the notfications array', () => {
    const container = (
      <IntlProvider defaultLocale="en" locale="en" messages={translation}>
        <Notification token="thisisatoken" />
      </IntlProvider>
    );
    const wrapper = mount(container).childAt(0);
    wrapper.setState({
      notifications: [{
        id: 'wewdwfed', content: 'this is the first comment', status: 'unread', createdAt: '2019-04-06',
      }],
    });
    wrapper.instance().setNewNotif(JSON.stringify({
      id: 'wewdwd', content: 'this is another comment', status: 'read', createdAt: '2019-04-04',
    }));
    const { state } = wrapper.instance();
    expect(state.notifications[0]).toEqual({
      id: 'wewdwd', content: 'this is another comment', status: 'read', createdAt: '2019-04-04',
    });
  });
  it('test setError', () => {
    const container = (
      <IntlProvider defaultLocale="en" locale="en" messages={translation}>
        <Notification token="thisisatoken" />
      </IntlProvider>
    );
    const wrapper = mount(container).childAt(0);
    wrapper.setState({ error: true });
    wrapper.instance().setError();
    const { state } = wrapper.instance();
    expect(state.error).toBe(false);
  });
});
