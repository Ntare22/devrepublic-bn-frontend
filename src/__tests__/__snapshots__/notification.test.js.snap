// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notification> render correctly date component 1`] = `
<div
  className="notification-wrapper"
>
  <button
    id="bell-button"
    onClick={[Function]}
    type="button"
  >
    <img
      alt="bell"
      className="notification-bell"
      src="test-file-stub"
    />
  </button>
  <div
    className="notif-panel container"
    style={
      Object {
        "display": "none ",
        "height": "280px",
      }
    }
  >
    <p>
      Notifications
      <span
        style={
          Object {
            "display": "none",
          }
        }
      >
        all notification already seen
      </span>
    </p>
    <div
      className="notifications"
      style={null}
    />
    <div
      className="row notif-buttons"
    >
      <button
        className="col"
        id="mark-all"
        onClick={[Function]}
        type="button"
      >
        mark all as read
      </button>
      <button
        className="col"
        id="view-more"
        onClick={[Function]}
        type="button"
      >
        view more
      </button>
    </div>
  </div>
</div>
`;
